Add client user guide for smithy java (#2522)
smithy-lang/smithy: docs/source-2.0/java/client/codegen-integrations.rst

Waiter Codegen
--------------

.. warning::

    Only synchronous waiters are supported at this time.

:ref:`Waiters <waiters>` are a client-side abstraction used to poll a resource until a desired state is reached,
or until it is determined that the resource will never enter a desirable end state.
Waiters can be defined in the Smithy model using the ``smithy.waiters#waitable`` trait.

For example:

.. code-block:: smithy
   :caption: model.smithy

    use smithy.waiters#waitable

    @waitable(
        BucketExists: {
            documentation: "Wait until a bucket exists"
            acceptors: [
                {
                    state: "success"
                    matcher: {
                        success: true
                    }
                }
                {
                    state: "retry"
                    matcher: {
                        errorType: "NotFound"
                    }
                }
            ]
        }
    )
    operation HeadBucket {
        input: HeadBucketInput
        output: HeadBucketOutput
        errors: [NotFound]
    }

The waiter-codegen integration can be used to automatically generate waiters from waitable trait

> definitions in your Smithy model. To add the integration to your project (using the Smithy Gradle plugins):

Reviewer: Using the waiters integration, you can automatically generate waiters from instances of the waitable trait in your Smithy model. If you are using the Smithy Gradle plugins, you can add this integration to your project like so:

> The waiter-codegen integration can be used to automatically generate waiters from waitable trait
> definitions in your Smithy model. To add the integration to your project (using the Smithy Gradle plugins):

Reviewer: ```suggestion
Using the waiters integration, you can automatically generate waiters from instances of the waitable trait in your Smithy model. If you are using the Smithy Gradle plugins, you can add this integration to your project like so:
```

Reviewer: Link to gradle plugins?

.. code-block:: kotlin
   :caption: build.gradle.kts

    dependencies {
        // Add codegen integration as a smithy-build dependency so it can be
        // discovered by the client codegen plugin
        smithyBuild("software.amazon.smithy.java.codegen:waiters:__smithy_java_version__")

        // Add waiters core package as a runtime dependency
        implementation("software.amazon.smithy.java:waiters:__smithy_java_version__")
    }

> This will cause your client code generator to create a waiter container class call ``<ServiceName>Waiters``
> in your generated client package. This container provides a method per waiter defined in your smithy model
> that returns a pre-configured ``Waiter``` instance base on the configuration set in your Smithy model.

Reviewer: ```suggestion
The code generator will create a class named ``<ServiceName>Waiters`` in the client package.
This class provides a method per waiter defined in your smithy model. The methods return ``Waiter``` instances based on the configuration set in your Smithy model.
```

> You can get an instance of this waiter container for a client by calling the ``waiters()`` method added
> to clients by this integration.

Reviewer: ```suggestion
To get an instance of a waiter, call the ``waiters()`` method on the client object:
```

Reviewer: Also add a simple explanation of the `wait()` method.

--------------
Waiter codegen
--------------

.. warning::

    Only synchronous waiters are supported at this time.

:ref:`Waiters <waiters>` are a client-side abstraction used to poll a resource until a
desired state is reached, or until it is determined that the resource will never enter
a desirable end state. Waiters can be defined in the Smithy model using the
``smithy.waiters#waitable`` trait.

For example:

.. code-block:: smithy
   :caption: model.smithy

    use smithy.waiters#waitable

    @waitable(
        BucketExists: {
            documentation: "Wait until a bucket exists"
            acceptors: [
                {
                    state: "success"
                    matcher: {
                        success: true
                    }
                }
                {
                    state: "retry"
                    matcher: {
                        errorType: "NotFound"
                    }
                }
            ]
        }
    )
    operation HeadBucket {
        input: HeadBucketInput
        output: HeadBucketOutput
        errors: [NotFound]
    }

Using the waiters integration, you can automatically generate waiters from instances

> of the waitable trait in your Smithy model. If you are using the Smithy Gradle plugins,

Author: ```suggestion
of the ``waitable`` trait in your Smithy model. If you are using the Smithy Gradle plugins,
```
